### PR TITLE
Finish Chapter1

### PR DESCRIPTION
**Pull request by oren-l**
### Solutions for Chapter 1

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter1.hs [hunk 1]
```diff
@@ -428,7 +428,7 @@ task is to specify the type of this function.
 >>> squareSum 3 4
 49
 -}
-
+squareSum :: Num a => a -> a -> a
```
Comment (vrom911) [inline src/Chapter1.hs:431]: We explain `Num` in the following chapters, so you are one step ahead 😸

In here `Int -> Int -> Int` would also work, as a more specific type 🙂

File: src/Chapter1.hs [hunk 2]
```diff
@@ -490,7 +490,8 @@ Implement a function that returns the last digit of a given number.
   whether it works for you!
 -}
 -- DON'T FORGET TO SPECIFY THE TYPE IN HERE
-lastDigit n = error "lastDigit: Not implemented!"
+lastDigit :: Integral a => a -> a
+lastDigit n = (abs n) `mod` 10
```
Comment (vrom911) [inline src/Chapter1.hs:494]: Nice work on using infix form of calling functions! We don't explain this syntax in our course to make it short. But well done :+1: 

Also, you can remove `()`, they are redundant in this case :slightly_smiling_face: 
```suggestion
lastDigit n = abs n `mod` 10
```

File: src/Chapter1.hs [hunk 3]
```diff
@@ -520,7 +521,9 @@ branches because it is an expression and it must always return some value.
   satisfying the check will be returned and, therefore, evaluated.
 -}
 closestToZero :: Int -> Int -> Int
-closestToZero x y = error "closestToZero: not implemented!"
+closestToZero x y = if abs(x) < abs(y)
```
Comment (vrom911) [inline src/Chapter1.hs:524]: No need to add brackets to one variable 🙂 

Also, in Haskell, all functions are separated from their arguments by space 👾 

```suggestion
closestToZero x y = if abs x < abs y
```

File: src/Chapter1.hs [hunk 4]
```diff
@@ -553,8 +556,11 @@ value after "=" where the condition is true.
 
 Casual reminder about adding top-level type signatures for all functions :)
 -}
-
-mid x y z = error "mid: not implemented!"
+mid :: Ord a => a -> a -> a -> a
+mid x y z
+  | x >= min y z && x <= max y z  = x
+  | y >= min x z && y <= max x z  = y
+  | otherwise                     = z
```
Comment (vrom911) [inline src/Chapter1.hs:563]: Very elegant solution 💃🏼

File: src/Chapter1.hs [hunk 5]
```diff
@@ -568,7 +574,8 @@ True
 >>> isVowel 'x'
 False
 -}
-isVowel c = error "isVowel: not implemented!"
+isVowel :: Char -> Bool
+isVowel c = any (==c) "aeiou"
```
Comment (vrom911) [inline src/Chapter1.hs:578]: Great 💡 
```suggestion
isVowel c = any (== c) "aeiou"
```

File: src/Chapter1.hs [hunk 6]
```diff
@@ -631,8 +638,10 @@ Implement a function that returns the sum of the last two digits of a number.
 Try to introduce variables in this task (either with let-in or where) to avoid
 specifying complex expressions.
 -}
-
-sumLast2 n = error "sumLast2: Not implemented!"
+sumLast2 :: Integral a => a -> a
+sumLast2 n = 
+  let last2 = (abs n) `mod` 100
```
Comment (vrom911) [inline src/Chapter1.hs:643]:
```suggestion
  let last2 = abs n `mod` 100
```

File: src/Chapter1.hs [hunk 7]
```diff
@@ -631,8 +638,10 @@ Implement a function that returns the sum of the last two digits of a number.
...
+  in  last2 `div` 10 + last2 `mod` 10
```
Comment (vrom911) [inline src/Chapter1.hs:644]: That is a wonderful solution! 👏🏼  You correctly noticed that it is the `div` and `mod`, cool 😎 

One hint to make your solution even shorter: you can see that you use both:

```haskell
mod ... 10
div ... 10
```
The standard library has the [`divMod`](https://hackage.haskell.org/package/base-4.14.0.0/docs/Prelude.html#v:divMod) function, that actually combines inside both `div` and `mod`. And this is exactly what you use!.

So you could write it this way:

```haskell
(a, b) = divMod ... 10
```

You can see how we could pattern match on the pair 🙂

File: src/Chapter1.hs [hunk 8]
```diff
@@ -652,8 +661,12 @@ Implement a function that returns the first digit of a given number.
 You need to use recursion in this task. Feel free to return to it later, if you
 aren't ready for this boss yet!
 -}
-
-firstDigit n = error "firstDigit: Not implemented!"
+firstDigit :: Integral a => a -> a
+firstDigit n = helper $ abs n
```
Comment (vrom911) [inline src/Chapter1.hs:665]: Fantastic idea to use a helper function 💎

**Review by vrom911 — APPROVED** [review]
Congrats on finishing Chapter 1! 🎉 
Awesome start 👏🏼

**Comment by chshersh** [comment]
We can't merge solutions to our repository, so I'm closing this PR.
Feel free to open PRs to unsolved chapters in your fork of this repository 🍴

**Pull request closed without merging**